File: packages/mcp-core/transport.ts
import { StdioServerTransport } from "@modelcontextprotocol/sdk/server/stdio.js";
import { Server } from "@modelcontextprotocol/sdk/server/index.js";
import type { MCPTransport } from "./types.ts";

/**
 * Standard stdio transport for MCP servers
 */
export class StdioTransport implements MCPTransport {
  private transport: StdioServerTransport;
  private server?: Server;
  
  constructor() {
    this.transport = new StdioServerTransport();
  }
  
  async connect(server?: Server): Promise<void> {
    if (server) {
      this.server = server;
      await this.server.connect(this.transport);
    }
  }
  
  async close(): Promise<void> {
    if (this.server) {
      await this.server.close();
    }
  }
}

/**
 * Socket transport for MCP servers (for testing/development)
 */
export class SocketTransport implements MCPTransport {
  constructor(private port: number) {}
  
  async connect(): Promise<void> {
    // Implementation for socket-based transport
    console.log(`Socket transport would connect on port ${this.port}`);
  }
  
  async close(): Promise<void> {
    console.log("Socket transport closing");
  }
}

File: packages/core/core/transport.py
"""
Transport implementations for MCP servers
"""

import logging
from typing import Any

from mcp.server import Server
from mcp.server.stdio import stdio_server

logger = logging.getLogger(__name__)


class StdioTransport:
    """
    Standard stdio transport for MCP servers
    """

    def __init__(self) -> None:
        self.server: Server[Any] | None = None

    async def connect(self, server: Server[Any] | None = None) -> None:
        """Connect the transport with an MCP server"""
        if server:
            self.server = server
            # Run the stdio server
            async with stdio_server() as (read_stream, write_stream):
                await self.server.run(
                    read_stream,
                    write_stream,
                    self.server.create_initialization_options(),
                )

    async def close(self) -> None:
        """Close the transport"""
        if self.server:
            logger.info("Stdio transport closing")


class SocketTransport:
    """
    Socket transport for MCP servers (for testing/development)
    """

    def __init__(self, port: int) -> None:
        self.port = port
        self.server: Server[Any] | None = None

    async def connect(self, server: Server[Any] | None = None) -> None:
        """Connect the transport on the specified port"""
        if server:
            self.server = server
        logger.info(f"Socket transport would connect on port {self.port}")

    async def close(self) -> None:
        """Close the socket transport"""
        logger.info("Socket transport closing")
